match2/commons/match_group_display_bracket.lua

```diff
@@ -1,738 +1,745 @@
+        opponentHeight = propsConfig.opponentHeight or defaultConfig.opponentHeight,
+        qualifiedHeader = propsConfig.qualifiedHeader or defaultConfig.qualifiedHeader,
+        roundHorizontalMargin = propsConfig.roundHorizontalMargin or defaultConfig.roundHorizontalMargin,
+        scoreWidth = propsConfig.scoreWidth or defaultConfig.scoreWidth,
+    }
 
-local _frame
-local _matches
+    local layoutsByMatchId, headMatchIds = Bracket.computeBracketLayout(props.matchesById, config)
+
+    local bracketNode = html.create('div'):addClass('brkts-bracket')
+        :css('--match-width', config.matchWidth .. 'px')
+        :css('--match-width-mobile', config.matchWidthMobile .. 'px')
+        :css('--score-width', config.scoreWidth .. 'px')
+        :css('--round-horizontal-margin', config.roundHorizontalMargin .. 'px')
+
+    for _, matchId in ipairs(headMatchIds) do
+        local nodeProps = {
+            config = config,
+            layoutsByMatchId = layoutsByMatchId,
+            matchId = matchId,
+            matchesById = props.matchesById,
+        }
+        bracketNode
+            :node(Bracket.NodeHeader(nodeProps))
+            :node(Bracket.NodeBody(nodeProps))
+    end
 
-function Bracket.get(frame)
-    local args = getArgs(frame)
-    return Bracket.luaGet(frame, args)
+    return html.create('div'):addClass('brkts-main brkts-main-dev')
+        :node(bracketNode)
 end
 
-function Bracket.luaGet(frame, args)
-    _frame = frame
-    
-    local bracketid = args[1]
-    local matches = DisplayHelper.getMatches(bracketid)
-    
-    if (args.matchHeight or '') == '' then
-        for id, match in utils.iter.spairs(matches) do
-            args.matchHeight = WikiSpecific.get_matchHeight(match.match2opponents[1] or {}, match.match2opponents[1] or {}, args.matchHeight)
-        end
-    end
-
-    local has3rd = false
-    for id, match in utils.iter.spairs(matches) do
-        if string.find(match.match2id or '', 'RxMTP') then
-            has3rd = true 
-            break
+Bracket.types.Layout = TypeUtil.struct({
+    height = 'number',
+    lowerMarginTop = 'number',
+    matchHeight = 'number',
+    matchMarginTop = 'number',
+    mid = 'number',
+    showHeader = 'boolean',
+})
+
+--[[
+Computes certain layout properties of nodes in the bracket tree.
+]]
+function Bracket.computeBracketLayout(matchesById, config)
+    -- Map match ids to their upper round matches
+    local upperMatchIds = {}
+    for matchId, match in pairs(matchesById) do
+        for _, x in ipairs(match.bracketData.lowerMatches) do
+            upperMatchIds[x.matchId] = matchId
         end
     end
 
-    _bracketConfig = _getBracketConfig(args)
+    -- Computes the layout of a match and everything to its left.
+    local computeNodeLayout = FnUtil.memoizeY(function(matchId, computeNodeLayout)
+        local match = matchesById[matchId]
+        local lowerLayouts = Array.map(
```

Comment:
Inconsistent indenting (continues from here)

Reply:
`computeNodeLayout` opens a closure - so it's indented one further